pkg/query-service/app/opamp/configure_ingestionRules.go

```diff
@@ -0,0 +1,228 @@
+			if err != nil {
+				zap.S().Error("failed to prepare OTLP internal receiver spec for agent", agent.ID, err)
+				return confHash, err
+			}
+			if err := agentConf.Merge(serviceConf); err != nil {
+				zap.S().Error("failed to merge OTLP internal receiver spec for agent", agent.ID, err)
+				return confHash, err
+			}
+		}
+	}
+
+	// add ingestion control
+	serviceConf, err := prepareControlProcesorsSpec(agentConf, Signal(signal), processors)
+	if err != nil {
+		zap.S().Error("failed to prepare ingestion control processors for agent ", agent.ID, err)
+		return confHash, err
+	}
+
+	err = agentConf.Merge(serviceConf)
+	if err != nil {
+		return confHash, err
+	}
+
+	// ------ complete adding processor
+	configR, err := yaml.Parser().Marshal(agentConf.ToStringMap())
+	if err != nil {
+		return confHash, err
+	}
+
+	zap.S().Debugf("sending new config", string(configR))
+	hash := sha256.New()
+	_, err = hash.Write(configR)
+	if err != nil {
+		return confHash, err
+	}
+	confHash = string(hash.Sum(nil))
+	agent.EffectiveConfig = string(configR)
+	err = agent.Upsert()
+	if err != nil {
+		return confHash, err
+	}
+
+	agent.SendToAgent(&protobufs.ServerToAgent{
+		RemoteConfig: &protobufs.AgentRemoteConfig{
+			Config: &protobufs.AgentConfigMap{
+				ConfigMap: map[string]*protobufs.AgentConfigFile{
+					"collector.yaml": {
+						Body:        configR,
+						ContentType: "application/x-yaml",
+					},
+				},
+			},
+			ConfigHash: []byte(confHash),
+		},
+	})
+
+	return string(confHash), nil
+}
+
+// prepare spec to introduce ingestion control in agent conf
+func prepareControlProcesorsSpec(agentConf *confmap.Conf, signal Signal, processors map[string]interface{}) (*confmap.Conf, error) {
+	agentProcessors := agentConf.Get("processors").(map[string]interface{})
+
+	for key, params := range processors {
+		agentProcessors[key] = params
+	}
+
+	updatedProcessors := map[string]interface{}{
+		"processors": agentProcessors,
+	}
+
+	updatedProcessorConf := confmap.NewFromStringMap(updatedProcessors)
+
+	// upsert changed processor parameters in config
+	err := agentConf.Merge(updatedProcessorConf)
+	if err != nil {
+		return agentConf, err
+	}
+
+	// edit pipeline if processor is missing
+	service := agentConf.Get("service")
+
+	signalSpec := service.(map[string]interface{})["pipelines"].(map[string]interface{})[string(signal)]
+	currentPipeline := signalSpec.(map[string]interface{})["processors"].([]interface{})
+
+	// merge tracesPipelinePlan with current pipeline
+	mergedPipeline, err := buildPipeline(signal, currentPipeline)
+	if err != nil {
+		zap.S().Error("failed to build pipeline", signal, err)
+		return agentConf, err
+	}
+
+	// add merged pipeline to the service
+	// todo(): check if this overwrites other processors
+	serviceConf := map[string]interface{}{
+		"service": map[string]interface{}{
+			"pipelines": map[string]interface{}{
+				string(signal): map[string]interface{}{
+					"processors": mergedPipeline,
+				},
+			},
+		},
+	}
```

Comment:
This is hard to follow/maintain with all the `map[string]interface{}`. Let's use the config updater. It would have been if this was occasionally used, but it is everywhere.

Reply:
Yes. I am going to replace it with actual params. 
I am not keen on getting config updater merged. It introduces unnecessary dependencies. It is going to be even harder to maintain.